### PR TITLE
armv7-a/armv7-r:cache: modify hardcode in cache set/way operation

### PR DESCRIPTION
**Pull request by zhuyanlinzyl**
Some chip not use the default cache size & way , read from
CCSIDR instead of hardcode.

Signed-off-by: zhuyanlin <zhuyanlin1@xiaomi.com>

## Testing
Test on qemu sabre-6quad:smp

### CODE REVIEW

**Review by pkarashchenko — APPROVED** [review]

**Comment by masayuki2009** [comment]
>Test on qemu sabre-6quad:smp

@zhuyanlinzyl 

Did you test this PR with your dual Cortex-A7 platform?

**Comment by zhuyanlinzyl** [comment]
Yes, I also have test on dual Cortex-A7 platform @masayuki2009

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@zhuyanlinzyl please rebase the latest mainline which fix the build break generated by other PR.

File: arch/arm/src/armv7-a/cp15_clean_dcache_all.S [hunk 1]
```diff
@@ -91,19 +91,30 @@
...
 
 	ldr		r3, =0x3ff			/* Isolate the way field (bits 3-12) */
 	and		r1, r3, r1, lsr #3		/* r1=(number of ways - 1) */
 
+	add		r4, r1, #1
+	clz		r4, r4
+	add		r4, #1				/* r4=Set/way operation Way shift */
+
+	ldr		r3,=0x7				/* Isolate the LineSize field (bits 0-2) */
+	and		r5, r3				/* r4=(Log2LineSize - 2) in word */
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-a/cp15_clean_dcache_all.S:110]: change to:
```
and r5, r3, r1
```
and remove line 97.

Comment (xiaoxiang781216) [reply arch/arm/src/armv7-a/cp15_clean_dcache_all.S:110]: apply the similar change to other place too.

Comment (zhuyanlinzyl) [reply arch/arm/src/armv7-a/cp15_clean_dcache_all.S]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@zhuyanlinzyl 
I noticed that sabre-6quad:smp with the dev board (not QEMU) has the following issue with this PR.
If I revert this PR, the issue does not happen.

```
nsh> uname -a
NuttX 10.2.0 5af1b671b6 Mar 29 2022 15:43:46 arm sabre-6quad
nsh> ps
  PID GROUP CPU PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK   STACK   USED  FILLED COMMAND
    0     0   0   0 FIFO     Kthread N-- Assigned           00000000 001000 000448  44.8%  CPU0 IDLE
    1     1   1   0 FIFO     Kthread N-- Running            00000000 001000 000208  20.8%  CPU1 IDLE
    2     2   2   0 FIFO     Kthread N-- Running            00000000 001000 000208  20.8%  CPU2 IDLE
    3     3   3   0 FIFO     Kthread N-- Running            00000000 001000 000208  20.8%  CPU3 IDLE
    4     4 --- 192 RR       Kthread --- Waiting  Semaphore 00000000 002016 000268  13.2%  hpwork 0x10821b54
    5     5   0 100 RR       Task    --- Running            00000000 002024 001124  55.5%  nsh_main
nsh> free
                   total       used       free    largest  nused  nfree
        Umem: 1065206176      12032 1065194144 1065194144     49      1
nsh> smp
  Main[0]: Running on CPU0
  Main[0]: Initializing barrier
Thread[1]: Started
  Main[0]: T

```

**Comment by zhuyanlinzyl** [comment]
hi @masayuki2009   

Thanks,  I'will investigate later. I will test in sabre-6quad:smp board.

**Comment by xiaoxiang781216** [comment]
@masayuki2009 we has ordered sabre-6quad early of this week, but will receive the board until the end of next week. So please be patient:(.

**Comment by zhuyanlinzyl** [comment]
hi @masayuki2009 :

I have a sabre-6quad board and use it to investigate this problem.
The board name is `Freescale i.MX6 Quad SABRE Smart Device Board`, and I use `sabre-6quad:smp` config.

I use tftp method and successful download nuttx bin

but the nuttx can run at all

```
=> setenv ipaddr 192.168.10.102
=> setenv serverip 192.168.10.103
=> tftp 0x10800000 nuttx.bin
Using ethernet@02188000 device
TFTP from server 192.168.10.103; our IP address is 192.168.10.102
Filename 'nuttx.bin'.
Load address: 0x10800000
Loading: #########
         9.4 MiB/s
done
Bytes transferred = 128144 (1f490 hex)
=> go 0x10800000
## Starting application at 0x10800000 ...

```

It hang in there and not any other log print.

Can you give me some advice?

**Comment by zhuyanlinzyl** [comment]
@masayuki2009  Is the loadaddr(0x10800000) error ? I just use same address from you

**Comment by masayuki2009** [comment]
@zhuyanlinzyl 

>It hang in there and not any other log print.
>
>Can you give me some advice?

How about enabling debug options and also disables ram log as follows.

```
--- a/boards/arm/imx6/sabre-6quad/configs/smp/defconfig 
+++ b/boards/arm/imx6/sabre-6quad/configs/smp/defconfig 
@@ -23,6 +23,9 @@ CONFIG_ARMV7A_WAYSIZE_64KB=y 
 CONFIG_BOARD_LOOPSPERMSEC=99369 
 CONFIG_BOOT_RUNFROMSDRAM=y 
 CONFIG_BUILTIN=y 
+CONFIG_DEBUG_ASSERTIONS=y 
+CONFIG_DEBUG_ERROR=y 
+CONFIG_DEBUG_FEATURES=y 
 CONFIG_DEBUG_FULLOPT=y 
 CONFIG_DEBUG_SYMBOLS=y 
 CONFIG_DEV_ZERO=y 
@@ -42,9 +45,6 @@ CONFIG_NSH_READLINE=y 
 CONFIG_PL310_LOCKDOWN_BY_LINE=y 
 CONFIG_PL310_LOCKDOWN_BY_MASTER=y 
 CONFIG_PREALLOC_TIMERS=4 
-CONFIG_RAMLOG=y 
-CONFIG_RAMLOG_BUFSIZE=16384 
-CONFIG_RAMLOG_SYSLOG=y 
 CONFIG_RAM_SIZE=1073741824 
 CONFIG_RAM_START=0x10000000 
 CONFIG_RAM_VSTART=0x10000000 
@@ -58,6 +58,7 @@ CONFIG_STACK_COLORATION=y 
 CONFIG_START_MONTH=3 
 CONFIG_START_YEAR=2016 
 CONFIG_SYMTAB_ORDEREDBYNAME=y 
+CONFIG_SYSLOG_TIMESTAMP=y 
 CONFIG_SYSTEM_NSH=y 
 CONFIG_SYSTEM_SYSTEM=y 
 CONFIG_SYSTEM_TASKSET=y 
```

>@masayuki2009 Is the loadaddr(0x10800000) error ? I just use same address from you

Yes, the address is correct.

**Comment by zhuyanlinzyl** [comment]
> @zhuyanlinzyl
> 
> > It hang in there and not any other log print.
> > Can you give me some advice?
> 
> How about enabling debug options and also disables ram log as follows.
> 
> ```
> --- a/boards/arm/imx6/sabre-6quad/configs/smp/defconfig 
> +++ b/boards/arm/imx6/sabre-6quad/configs/smp/defconfig 
> @@ -23,6 +23,9 @@ CONFIG_ARMV7A_WAYSIZE_64KB=y 
>  CONFIG_BOARD_LOOPSPERMSEC=99369 
>  CONFIG_BOOT_RUNFROMSDRAM=y 
>  CONFIG_BUILTIN=y 
> +CONFIG_DEBUG_ASSERTIONS=y 
> +CONFIG_DEBUG_ERROR=y 
> +CONFIG_DEBUG_FEATURES=y 
>  CONFIG_DEBUG_FULLOPT=y 
>  CONFIG_DEBUG_SYMBOLS=y 
>  CONFIG_DEV_ZERO=y 
> @@ -42,9 +45,6 @@ CONFIG_NSH_READLINE=y 
>  CONFIG_PL310_LOCKDOWN_BY_LINE=y 
>  CONFIG_PL310_LOCKDOWN_BY_MASTER=y 
>  CONFIG_PREALLOC_TIMERS=4 
> -CONFIG_RAMLOG=y 
> -CONFIG_RAMLOG_BUFSIZE=16384 
> -CONFIG_RAMLOG_SYSLOG=y 
>  CONFIG_RAM_SIZE=1073741824 
>  CONFIG_RAM_START=0x10000000 
>  CONFIG_RAM_VSTART=0x10000000 
> @@ -58,6 +58,7 @@ CONFIG_STACK_COLORATION=y 
>  CONFIG_START_MONTH=3 
>  CONFIG_START_YEAR=2016 
>  CONFIG_SYMTAB_ORDEREDBYNAME=y 
> +CONFIG_SYSLOG_TIMESTAMP=y 
>  CONFIG_SYSTEM_NSH=y 
>  CONFIG_SYSTEM_SYSTEM=y 
>  CONFIG_SYSTEM_TASKSET=y 
> ```
> 
> > @masayuki2009 Is the loadaddr(0x10800000) error ? I just use same address from you
> 
> Yes, the address is correct.

I open the debug config, but still no other log print after `## Starting application at 0x10800000 ...`
I also compare the image in board with my PC build.
No change in bin difference.

```
=> md 0x10800000
10800000: e59ff018 e59ff018 e59ff018 e59ff018    ................
10800010: e59ff018 e59ff018 e59ff018 e59ff018    ................
10800020: 10800040 108004c0 10800280 10800400    @...............
10800030: 10800340 10800568 108001c0 10800560    @...h.......`...
10800040: e3a000df e121f000 ee110f10 e3c00005    ......!.........
10800050: e3c00a01 ee010f10 e59f50d8 e1a00005    .........P......
10800060: e3a01000 e2802901 e4801004 e4801004    .....)..........
10800070: e4801004 e4801004 e1300002 1afffff9    ..........0.....
10800080: e28f00b4 e890000e e1810003 e7850922    ............"...
10800090: e28f00b0 e890001e e0852922 e1810003    ........")......
108000a0: e2544001 e4820004 e2811601 1afffffa    .@T.............
108000b0: e3a00000 ee080f17 ee080f16 ee080f15    ................
108000c0: ee070fd5 ee070f15 e3851048 ee021f10    ........H.......
```

Do you have any other advice? @masayuki2009

**Comment by zhuyanlinzyl** [comment]
hi, @masayuki2009 
I found error in cp15_invalidate_dcache_all.
I fix this bug in https://github.com/apache/incubator-nuttx/pull/6107
I check success. You can try it

**Pull request merged**